hazelcast/src/main/java/com/hazelcast/client/impl/connection/tcp/TcpClientConnectionManager.java

```diff
@@ -462,8 +467,20 @@ Connection connect(Object target, Function<Object, Connection> getOrConnectFunct
         } catch (ClientNotAllowedInClusterException e) {
             logger.warning("Exception during initial connection to " + target + ": " + e);
             throw e;
+        } catch (TargetDisconnectedException e) {
+            logger.warning("Exception during initial connection to " + target + ": " + e);
+            connectionProcessListener.remoteClosedConnection(addressProvider.apply(target));
+            return null;
         } catch (Exception e) {
             logger.warning("Exception during initial connection to " + target + ": " + e);
+            if (e.getCause() instanceof IOException) {
+                try {
+                    connectionProcessListener.connectionAttemptFailed(addressProvider.apply(target));
+                } catch (Exception e2) {
+                    logger.warning("failed to translate address, can't fire connectionAttemptFailed() event for target"
+                            + target, e2);
+                }
+            }
```

Comment:
I am going to repeat myself, but I am not satisfied why we just call `connectionAttemptFailed` only when the cause is `IOException`. Right now, the connection attempt clearly failed one way or another. Why don't we call the listener?